### DISCUSSION TITLE
Extremely high CPU usage of the `terminal.draw` method

### DISCUSSION THREAD

**Discussion by gtema** [ratatui/ratatui#1927] [Discussion]
## Description
Ratatui applications are causing a very high CPU usage depending on the draw rate. For example https://github.com/ratatui/templates/tree/main/component-generated template with no changes (which draws static content at 60 FPS) is causing 50% utilization of a single core in debug build and still 7% in release build. Bisection lead me to the [`tui.draw`](https://github.com/ratatui/templates/blob/main/component-generated/src/app.rs#L166) call. Commenting out all operations inside the closure (invoking `tui.draw` with empty body) is not having any effect on the CPU, while commenting the `tui.draw` itself immediately drops CPU usage.

## To Reproduce

Just run the https://github.com/ratatui/templates/tree/main/component-generated app and see the CPU usage.

## Expected behavior

It is hard to say what is expected CPU usage, but I would say 1% CPU usage would be a great goal. Looking at TUI written in go (k9s) it does not use CPU so much even when you permanently cause actions resulting in a redraw, what hints there is something not really optimal done in the code.

Until there is a fix recommending a pattern of redrawing without the need is not helpful, so at least some change in the template that demonstrates how to only draw when necessary should be implemented since it is hard for people just looking first time at ratatui to figure out how to achieve this. On its own I would not name it a proper solution to only draw when necessary since even when redrawing at event (keypress) and active interaction you can still see the spikes and laptop battery is drained pretty fast

## Environment
- OS: Fedora 40
- Terminal Emulator: Konsole (kde)
- Font:
- Crate version: 0.27-0.28
- Backend: crossterm

**Comment by kdheepak** [comment]
Decreasing the frame rate is one way to decrease CPU usage:

https://github.com/ratatui/templates/blob/c23de5ee239ad98811c92ef1c4149fdc067e65c2/component-generated/src/tui.rs#L66-L67

**Comment by gtema** [comment]
I know where to reduce the rate (what is actually in the referred template better controlled on the cli defaults level), but that is only a workaround and not the solution. Even with the refresh rate at 0 and only render when you have a change (what the template does not show) you are still having much higher cpu usage per frame compared to TUI written in Golang (we all know Rust would have higher bare performance compared to go, even when not so much. So saying we can not have better or at least equal performance in Rust for TUI compared to Go is not a good sign). This is not a problem of laptop/terminal/application but rather the framework that is not efficient enough. 

I strongly believe there should be relatively small inefficiency in the code with the very high effect since few tiny changes in my app based on ratatui got extremely high cpu influence

**Comment by joshka** [comment]
Bumped the templates up to use the latest dependencies just now and ran
```shell
CARGO_PROFILE_RELEASE_DEBUG=true cargo flamegraph --root
```

![image](https://github.com/user-attachments/assets/e5e333f3-efef-47fd-a9d8-662fffe41ded)
zooming in
<img width="854" alt="image" src="https://github.com/user-attachments/assets/880bfba3-0617-43b1-a08b-dcf103fd5930">
<img width="653" alt="image" src="https://github.com/user-attachments/assets/74693ce2-91fc-4e4a-81f5-b2737ed11f64">

What you're seeing here is the terminal buffer diffing algorithm happening. Some context:
- Ratatui does not track per widget information between buffers
- Ratatui's buffer is a vec of cells containing strings (supporting unicode)
- The terminal has two buffers, one which was previously drawn and the current one which is yet to be drawn
- Actually writing to the terminal can be slow, and needs code to avoid doing that
- Ratatui avoids writing to the terminal by diffing the two buffers on each frame to work out the necessary changes to draw
- In situations like static text, the diffing algorithm dominates over any other code

One of the things that stands out immediately as a hotspot is that the diff calls `.width()` twice - this is an easy fix that gains 17%. There's likely another %20 to be easily had there too. See https://github.com/ratatui/ratatui/pull/1339 for details.

Taking a wider perspective on this, there's definitely some inefficiencies in the way that the symbols are stored, and the hence the algorithms used to speed up display are also inefficient. It's likely that storing symbols as runs of text rather than splitting them up into single cell strings could be a useful optimization. (This is an untested hypothesis though).

Looking at where the flamegraph seems to be dominated after the change mentioned above, points to the window size check being a problem. This might be an area where we can possibly keep state around instead of re-checking, however doing so would interact poorly with the user code calling into crossterm. This isn't something that would be easily possible due to Ratatui not being a library, not a framework (you call us, we don't call you).

Some things to explore to speed this up in general would be:
- making it possibly to turn off the diffing between frames and just rewrite the entire frame (this has been previously requested). This is a viable thing to explore.
- replacing the current cell oriented storage approach (this is a huge undertaking to get right in a way that would continue to work with existing code)
- adding metrics around the various parts of the rendering pipeline (this is something I'm already exploring) to get more general insight in real apps about the parts of the app that are actually causing performance issues.
- looking at whether any of the tasks can be parallelized in a way that cuts down the CPU wall time
- caching more of the linear calculations (like the symbol width) between frames
- switching to something faster than unicode-width (not sure if there are any specific bottlenecks here)
- making it possible for widgets to provide hints about what's changed (not sure this is a good idea)
- adding a more framework style integration with crossterm so that we don't spend time on window size checks (and instead react to events as they come in). This is likely not possible
- It's possible that doing partial rendering instead of full rendering would be a useful approach (but also a fundamentally changing one).
- It's possible that there's some way to track buffer writes in a way that would help pre-calculate the diff when drawing (e.g. tracking cell changes as they happen on each frame). This is likely very difficult and a fundamental change.

---

For the most part, making apps that redraw their screen when nothing has actually changed seems to be a bad idea to me. This practice kind of comes from the historical examples from tui-rs which all had calls to event::poll, and which redrew the frame at a particular rate. This is generally unnecessary, and apps should generally be designed such that they draw frames only when necessary. Redrawing at 60 FPS is useless if every frame is drawing the same thing. That said, 60 FPS is likely a good thing to take note of when there is animation or rapidly changing information in the app. But even there, 30 FPS or much less is likely good enough for most terminal apps.

**Comment by kdheepak** [comment]
We could introduce a single `dirty` boolean flag that is set to `false` after a `draw` but set to `true` in all `render` calls. And then only calculate the diff and output to the terminal when `dirty`. 

That said, drawing at 60FPS or 30FPS will still be slower than Go or any other library in any language if they don't also draw at 60FPS or 30FPS. Immediate mode rendering in the terminal like this is just always going to do more work.

**Comment by joshka** [comment]
> We could introduce a single `dirty` boolean flag that is set to `false` after a `draw` but set to `true` in all `render` calls. And then only calculate the diff and output to the terminal when `dirty`.

Do you mean In the template or in Ratatui? I'm not quite sure what you're proposing here.

> That said, drawing at 60FPS or 30FPS will still be slower than Go or any other library in any language if they don't also draw at 60FPS or 30FPS. Immediate mode rendering in the terminal like this is just always going to do more work.

I think you're saying here that rendering more frames per second is slower than rendering less frames right, or have I misunderstood this?

**Comment by kdheepak** [comment]
> Do you mean In the template or in Ratatui? I'm not quite sure what you're proposing here.

I was thinking in the template but if it can be added into the core library that'd work too. I'm not sure if it's possible though? 

Currently we have logic like this (pseudo code below) in the template like this:

```rust
while true {
   app.update();

   terminal.draw(|f| {
      f.render_widget(&app)
   });

   sleep(1/60)
}
```

But it could be written like this:

```rust
while true {
   app.update(); // sets `dirty` flag to `true` only when there are updates

   if app.dirty {
       terminal.draw(|f| {
          f.render_widget(&app)
       });
       app.dirty = false
    }

   sleep(1/60)
}
```

Without introducing an `App` like struct into the core library, I don't know how we'd introduce a `dirty` flag to the library and not break everyone's code.

**Comment by joshka** [comment]
Neither did I ;) I'd probably use a draw message rather than a dirty flag , but there are downsides to both approaches.

**Comment by orhun** [comment]
It would be best to leave the handling of `dirty` flag and other less internal rendering/performance improvements to the user. In the meantime, #1339 is a really good start for doing little but effective caching tweaks to gain more performance.

> Some things to explore to speed this up in general would be:

There seem to be a lot of areas to explore for significant improvements \o/

**Comment by JonasVautherin** [comment]
I think I have a similar situation: I wrote a small messenging TUI that allows the user to enter text. I only call `draw` when I know that there is a state change.

Whenever the user types a character, I need to redraw. Unless I miss something fundamental, this seems pretty obvious.

My process takes roughly 0% CPU when idling, but it easily goes up to 10% when I'm writing a text. It feels like this is a lot of CPU just to add a character to an otherwise completely static screen, right?

I made a flamegraph with the following command:

```
CARGO_PROFILE_RELEASE_DEBUG=true cargo flamegraph
```

and it seems like `draw` and `diff` take the most resources:

![Image](https://github.com/user-attachments/assets/090a90ae-aff8-4b7c-be6e-596ad587fe23)

Is there something I can do to optimise that? It is related to this issue, right?

**Comment by joshka** [comment]
> Is there something I can do to optimise that? It is related to this issue, right?

My generic answer to this is to follow a performance tuning checklist. Here's how I see that:

1. Identify specific measurable target performance goals that align to your app's needs (latency / throughput / utilization). These should be based on some real world impact, and avoid synthetic values. The goals should be expressed in units that are meaningful to the user.
2. Measure those characteristics with a profiler
3. Identity places where the goals aren't met
4. Identify how the bottlenecks that prevent that
5. Improve the bottlenecks
6. Add benchmarks to show that this is improved (and the perf can be tracked over time)
7. Re-measure
8. Repeat

When in doubt, JFMI (Just Measure It).

Generally, no step is valid in that process unless the previous step is properly complete, skipping any step often leads to solving the wrong problems and wasting peoples time.

The reality is that CPU usage doesn't matter unless it's causing actual problems (latency, cost, impact on other programs, battery life, etc.). The reality is that RAM usage doesn't matter unless it's causing actual problems for other programs. The reality is that latency in UI code doesn't matter unless it's at a level where it's meaningful to humans detecting that latency (order of magnitude 10-100ms level). Throughput matters if it's causing wait times in the seconds to minutes range.

Only optimize what matters.

All that said, given that the buffer diff algortihm is something known to be expensive, I'd be fine with a change that turns the diff off completely and instead re-renders the entire buffer on each frame. This would give a nice big lever to use to attack one of the larger parts of the per frame processing.

The other part where there's likely (based on some educated guesses rather than actually following the steps) to be perf gains is in the way that text is stored and split into cells. There's some inherent overhead in that approach over keeping everything as strings which can be written all at once. The problem with fixing this part is you need to have a deep understanding of the internals, and then you need to be able to fix it in a way that continues to be compatible with all the released software so that it doesn't cause a massive overhead to upgrade between each version for all users. And you need to consider the maintenance burden of "improving" that code (i.e. if you add complexity to help solve performance, then you need to commit the time to support that complexity over the indeterminate future - the upfront change cost is just a small part of the total effort).

**Comment by JonasVautherin** [comment]
I would be interested in understanding when the diff algorithm is useful vs when it is not, because just from this thread I get the sense that it is slowing things down. Surely the diff algorithm is there for a reason, isn't it?

> The reality is that CPU usage doesn't matter unless it's causing actual problems (latency, cost, impact on other programs, battery life, etc.). The reality is that RAM usage doesn't matter unless it's causing actual problems for other programs. The reality [...]

I am honestly not sure what to think of that. Using 10% CPU to print text to the screen doesn't sound like a premature optimization to me: it's just a lot worse than pretty much any way I know to sanely do this in all the languages I know 🤔.

Don't take it the wrong way: if this is a fundamental limitation of the framework, then so be it. I obviously don't expect you (or anyone) to fix it just so I can get better performance in my toy project 😇. I just find it surprising that it seems to be a philosophy. Why use Rust at all, in that case?

**Comment by kdheepak** [comment]
I think there's a few points here to think about:

1. If your app is not doing any work, i.e. you press a key and draw, the flamegraph is going to show lots of time spent in the `ratatui::draw` function, which is probably normal? "number of ticks per second", i.e. number of times you call the draw function per second can increase CPU utilization. For example, in https://github.com/kdheepak/taskwarrior-tui/issues/122 I had accidentally sent the default to end up with very high CPU usage. Are you able to share your code? That'll help identify what might be going on here.
2. When drawing, ratatui diff's the previous buffer and the current buffer and only renders the individual characters at the positions where they are different. In some tests of line, I found that if more than 30-40% of the characters were different, the time taken to calculate the diff was not worth it. This discussion has some useful information that's probably relevant - https://github.com/ratatui/ratatui/discussions/579
4. Looking at the flamegraph, it seems like your code spends equal amounts of time in `sircuit::draw` and in `ratatui::diff`. What's in `sircuit::draw` is the time taken to build a widget and render the widget. I think only way to speed that up is to build the widgets outside of the draw function (which has its own tradeoffs; see documentation for `WidgetRef`). I also see time being spent in unicode width compares in the `ratatui::diff`, which we can probably allow users to turn off if there are rendering pure ascii TUIs?

**Comment by joshka** [comment]
> Don't take it the wrong way: if this is a fundamental limitation of the framework, then so be it. I obviously don't expect you (or anyone) to fix it just so I can get better performance in my toy project 😇. I just find it surprising that it seems to be a philosophy. Why use Rust at all, in that case?

It's less a limitation and more a matter of prioritization. A TUI's performance shouldn't generally be in a hot path of anything which would cost you money (exception would be perhaps finance dashboards). And performance tuning UIs for single digit milliseconds is never going to be noticeable by a human. Performance tuning takes a rigorous amount of energy to do properly. That energy is often better spent elsewhere unless there's a payoff that makes the time spent worth it. I haven't found anyone to make a good argument for that so far (and would change my mind if they did).

The biggest hassle with working out whether the diff makes sense is working out representative changes. It's also sensitive to the specific terminal emulator you use, and I suspect hardware performance will make a huge amount of difference to how much benefit you'll see.

Here's benchmark output on my Mac M2 under iTerm2, where I'm just randomly changing a percentage of full lines.
Each line is replaced with random symbols and indexed fg/bg colors.
"With diff" includes the diff in the benchmark measurement
"Without diff" computes a diff from a buffer filled with placeholder symbols that would cause every line to be seen as different.

<img width="1708" alt="Image" src="https://github.com/user-attachments/assets/96ee0360-a496-4b8d-93fe-b5eccf781b89" />

Run https://github.com/ratatui/ratatui/pull/1593 (through its paces on your hardware (particularly if you have something lower powered that can show larger delta between diff and no diff. Experiment with the way the changes are made (lines, chars, styles, etc.)

One bounding goal to note is that this is single threaded (the diff could be put off thread, but it's not currently), and has a realistic time budget of 16.7ms if you want your UI to hit 60fps.

**Comment by joshka** [comment]
sorry - fat finger closed

**Comment by kdheepak** [comment]
Interesting benchmark! So with a higher percentage of changes between the buffers, there's a more chances of "slower" diffs, but the average seems to be the same. 

When I was testing this this was also very sensitive to the size of the terminal. 

>  I haven't found anyone to make a good argument for that so far (and would change my mind if they did).

For me the biggest argument is that there's an art / science to writing performant code and code that does zero allocations, it can show a mastery of the language. And it's a skill I'd personally like to develop. I also think people have come to expect performant / zero allocations in rust libraries. To that end, if there are obvious improvements to make that would benefit a majority (changes both in the core library / in templates etc) I think we should push to do that. 

That said, it can be a time consuming investment and we may be constrained by breaking changes for very little benefit, so it may not be worth it from that perspective.

**Comment by joshka** [comment]
IMO, the best first step I know of to starting on this  is to add metrics, so that rendering performance can be measured at runtime with real loads. Likely using `autometrics`.

**Comment by JonasVautherin** [comment]
> Are you able to share your code? That'll help identify what might be going on here.

Probably easiest looking at one of the official [examples: user_input](https://ratatui.rs/examples/apps/user_input/). If It only redraws when a key is pressed. In release mode (i.e. I run with `--release`), it takes ~0% CPU when idle, but 5% when I type a message. As a comparison, aerc stays at 0%, vi stays at 0%, zsh stays at 0% in the same conditions.

The "user_input" example has an "Input" area and a "Messages" area. If I add a second "Messages" area, it goes up to 7% CPU.

> And performance tuning UIs for single digit milliseconds is never going to be noticeable by a human.

I don't think anyone here is talking about tuning for single digit milliseconds 🤔. The issue is about high CPU usage...

> build the widgets outside of the draw function (which has its own tradeoffs; see documentation for WidgetRef

Oh thank you! I'll look into that! What are the tradeoffs (other than it is unstable?). It feels like it gives me the control over when I want to actually build the widget (so I can do it only when I know it changes), but I don't see any downside yet 😅. I guess it still won't run `diff` outside the draw function, though?

EDIT: actually I don't get what you mean with the `WidgetRef` from the documentation. It feels like it will still be rendered in the draw function (with `frame.render_widget_ref` instead of `frame.render_widget`) 🤔 🤔.

**Comment by joshka** [comment]
I'm moving this to a discussion, mainly because it's not really describing a fixable problem with changes to Ratatui.

The solution to doing less work is to choose to do less work. Here that's effectively:
- only render when there's updates (not at some FPS)
- don't rebuild the entire widget state every frame (render references to widgets instead of consuming them)

**Reply by JonasVautherin** [reply]
> only render when there's updates (not at some FPS)

Doesn't the fact that it uses 5% CPU when typing text but 0% when not typing text suggest that it doesn't render at some FPS, but rather it renders when there's updates (specifically, when a key is pressed and a letter needs to be printed)?

> don't rebuild the entire widget state every frame (render references to widgets instead of consuming them)

I wanted to do something like what Compose does, i.e. only redraw the part that changes. Could be that "references to widgets" are allowing that, but I haven't managed to make it work.

Maybe that's a simple one: how would you change the [user_input example](https://ratatui.rs/examples/apps/user_input/) in such a way that it doesn't take 5% CPU when typing? Or are you saying that it's not fixable as in, Ratatui is fundamentally not meant for such use-cases?
